### PR TITLE
[rc] install C++20 modules; enable tzdb integration

### PR DESCRIPTION
**Pull request by h-vetinari**
This addresses most of #142 (for the rc branch), by installing C++20 modules (enabled by default since https://github.com/llvm/llvm-project/commit/19d2d3fe50c301272350d12c53c801b17e29e64e) as well as pointing the experimental chrono implementation to use our own `tzdata`, following very closely along the lines already discussed in detail on the [libstdcxx side](https://github.com/conda-forge/ctng-compilers-feedstock/pull/142)

### CODE REVIEW

**Comment by h-vetinari** [comment]
@isuruf, you mentioned in https://github.com/conda-forge/ctng-compilers-feedstock/pull/142 that patching in the relative paths approach for the tzdb _in a static library_ would be a security hole (presumably because the symbol could be used in a way to load an arbitrary payload), but we have a problem here: libcxx 19 only has experimental tzdb-support, which ships in the [static](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0-rc1/libcxx/src/CMakeLists.txt#L358) `libc++experimental.a` (as the patch shows I was barking up the wrong tree here so far).

Given that libcxx 19 has [no native support](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0-rc1/libcxx/CMakeLists.txt#L117) for `<chrono>` on osx (and I believe this is an important usecase, not least because it's the last C++ stdlib to haven't completed support and projects are starting to use it), could you live with the environment variable approach here so that we at least have _some_ way to use the tzdb bits on osx? Once this moves into the shared library, I'd be happy to switch to relative paths.

**Comment by h-vetinari** [comment]
Hm, though I guess anything in a static library that needs to do some sort of dynamic look-up can be exploited, also the env var approach. I guess we could do something like GCC [does](https://github.com/gcc-mirror/gcc/blob/releases/gcc-14.1.0/libstdc%2B%2B-v3/src/c%2B%2B20/tzdb.cc#L1090-L1092) with `_GLIBCXX_STATIC_TZDATA`, i.e. embedding the tzdb into the static lib?

**Comment by isuruf** [comment]
In the gcc PR, we used the system tzdb for static library. We should do the same.

**Comment by h-vetinari** [comment]
Not sure I understand. The static variant of `zoneinfo_dir_override` does
```C
 #ifdef _GLIBCXX_ZONEINFO_DIR
     return _GLIBCXX_ZONEINFO_DIR;
 #else
     return nullptr;
 #endif
```
which is [defined](https://github.com/gcc-mirror/gcc/blob/releases/gcc-14.1.0/libstdc%2B%2B-v3/acinclude.m4#L5674) as `/usr/share/zoneinfo` on linux, but [undefined](https://github.com/gcc-mirror/gcc/blob/releases/gcc-14.1.0/libstdc%2B%2B-v3/acinclude.m4#L5684) on darwin (libcxx [comes](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0-rc1/libcxx/CMakeLists.txt#L107-L118) to the same conclusion). So what system tzdb do you mean?

**Comment by isuruf** [comment]
`/usr/share/zoneinfo` works just fine on macos (At least on macOS 14)

**Comment by h-vetinari** [comment]
> `/usr/share/zoneinfo` works just fine on macos (At least on macOS 14)

does it contain `tzdata.zi` and `leapseconds`, or just the zoneinfo folders? Actually, it seems libcxx intentionally bases itself on another [style](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0-rc1/libcxx/docs/DesignDocs/TimeZone.rst#leap-seconds) of leapseconds file, which we don't have in our `tzdata`. 😑

**Comment by h-vetinari** [comment]
Given the missing `leap-seconds.list` almost everywhere (and generally unclear support situation on older macs), can we go back to the idea about embedding `tzdata.zi` & `leap-seconds.list` statically?

That would keep the tzdb basically as up to date as explicitly depending on `tzdata` (because we release `libcxx` builds regularly, but `tzdata` barely ever updates), and the footprint is small (~110kb, less than 10% of the current size). If size is an issue, we could also put `libc++experimental.a` into a separate output (which might be a good idea anyway).

**Comment by isuruf** [comment]
I don't think embedding those are good. This is akin linking statically and we discourage that. I'd rather we wait until the tzdb integration matures enough to move out of `libc++experimental.a`.

**Comment by h-vetinari** [comment]
> I'd rather we wait until the tzdb integration matures enough to move out of `libc++experimental.a`.

That condemns everyone in conda-forge to wait likely at least a year just to use C++20 `<chrono>` (which is available on all other platforms) for what comes down to a matter of taste. Yes, we discourage static linking, but cfep-18 has exceptions for a reason. We'll move to the same approach as for libstdcxx as soon as it's in the shared lib, but until then, this case is easily important enough for an exception.

**Comment by isuruf** [comment]
> That condemns everyone in conda-forge to wait likely at least a year just to use C++20 <chrono>

I haven't heard from anyone about this. Are there any use-cases?

**Comment by isuruf** [comment]
What does other package managers on macOS do?

**Comment by h-vetinari** [comment]
There's at least [two](https://github.com/conda-forge/fimex-feedstock/blob/main/recipe/meta.yaml) [feedstocks](https://github.com/conda-forge/sparrow-feedstock/blob/main/recipe/meta.yaml) using `howardhinnant_date` as a replacement because they cannot use C++20 `<chrono>` (Howard authored [P0355](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0355r7.html) upon which it is based, and his [implementation](https://github.com/HowardHinnant/date) functions like a backport). I know that sparrow originally wanted to use C++20, because I got pinged about chrono support in conda-forge.

Looking around, onnxruntime similarly used [`libdate`](https://github.com/conda-forge/libdate-feedstock) (which is unfortunately identical to `howardhinnant_date`...), but now seems to [require](https://github.com/microsoft/onnxruntime/issues/21033) C++20 already based on the upstream issue tracker. There's also `r-rcppdate` which is yet another implementation of Howards code, adapted for R.

Other projects I'm aware of:
* arrow [vendors](https://github.com/apache/arrow/blob/main/cpp/src/arrow/vendored/datetime/tz.h) some form of Howard's code 
* orc [needs](https://github.com/apache/orc/blob/v2.0.1/c%2B%2B/src/Timezone.cc#L675) a tzdb already and [wants](https://github.com/apache/orc/issues/1577#issuecomment-1679879646) to switch to C++20
* qt [uses](https://github.com/qt/qtbase/blob/6.5.0/src/corelib/time/qdatetime.h#L464) the tzdb-bits already (though behind a config)

C++20 support is really on the uptake though (as implementations are finally starting to stabilize in gcc/clang), so I expect this to become even more relevant over the next 12 month, especially as libcxx is the last major implementation to ship tzdb-support (and since date-handling is so central to many use cases yet so painful to hand-roll, people have a big incentive to switch).

**Comment by h-vetinari** [comment]
> What does other package managers on macOS do?

I'm assuming newer apple versions will ship what they need in `/usr/share/zoneinfo`, since the libcxx BDFL works there. For other distribution, I don't know. I'm assuming many will just override `__libcpp_tzdb_directory`, which is why it's a weak symbol after all.

**Comment by h-vetinari** [comment]
I don't have access to a mac and `/usr/share/zoneinfo` is not part of the [SDKs](https://github.com/alexey-lysiuk/macos-sdk), but I guess if all the required files are there on a macOS 14 agent (including `leap-seconds.list`), then we could just document that the solution is to use `vmImage: macos-14`. That would at least be a way to make it work partially without impacting the deployment target.

**Comment by h-vetinari** [comment]
OK, so ubuntu-22.04 doesn't contain `leap-seconds.list`, but I checked that `ubuntu-24.04` does (however, using the beta images didn't work in azure for now). `macos-14` also doesn't contain `leap-seconds.list`, but does contain a working `tzdata.zi`. So I guess the lowest non-trivial support to have would be "timezones only". Pragmatically speaking, those are much more important than leapseconds, so I guess that's a start.

File: recipe/patches/0005-patch-__libcpp_tzdb_directory-to-allow-use-on-osx.patch [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+-endif()
+ option(LIBCXX_ENABLE_TIME_ZONE_DATABASE
+   "Whether to include support for time zones in the library. Disabling
+   time zone support can be useful when porting to platforms that don't
+   ship the IANA time zone database. When time zones are not supported,
+-  time zone support in <chrono> will be disabled." ${ENABLE_TIME_ZONE_DATABASE_DEFAULT})
++  time zone support in <chrono> will be disabled." ON)
+ option(LIBCXX_ENABLE_VENDOR_AVAILABILITY_ANNOTATIONS
+   "Whether to turn on vendor availability annotations on declarations that depend
+    on definitions in a shared library. By default, we assume that we're not building
```
Comment (isuruf) [inline recipe/patches/0005-patch-__libcpp_tzdb_directory-to-allow-use-on-osx.patch:35]: You don't need this part of the patch. You can just set `-DLIBCXX_ENABLE_TIME_ZONE_DATABASE=ON`

Comment (h-vetinari) [reply recipe/patches/0005-patch-__libcpp_tzdb_directory-to-allow-use-on-osx.patch:35]: Yeah, sure.

**Comment by h-vetinari** [comment]
OK, I've skipped the leapsecond portion, but now we fail in the override test; `_LIBCPP_WEAK` should be [set](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0-rc1/libcxx/include/__config#L800), and the libcxx test suite [does](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0-rc1/libcxx/test/libcxx/time/time.zone/time.zone.db/version.pass.cpp#L34) the override in pretty much the same way.

**Comment by h-vetinari** [comment]
OK, so libcxx doesn't seem to want external users to override that weak symbol. See [here](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0-rc1/libcxx/test/support/test_tzdb.h#L21-L23):
```C
// This function is marked as "overridable" in libc++ only for the test
// suite. Therefore the declaration is not in <chrono>.
```
So I think we should drop that test. Unless you want to patch libcxx to make their symbol externally overridable?

**Comment by h-vetinari** [comment]
> So I think we should drop that test. Unless you want to patch libcxx to make their symbol externally overridable?

I did that now; the rest passes. PTAL

**Comment by h-vetinari** [comment]
Going with this approach for now. We're still only in the rc-branch, so we'll have time to refine before going to `main`.

**Pull request merged**